api/v1/api.go

```diff
@@ -0,0 +1,212 @@
+	errorBadData     errorType = "bad_data"
+	errorInternal    errorType = "internal"
+	errorUnavailable errorType = "unavailable"
+	errorNotFound    errorType = "not_found"
+)
+
+type apiError struct {
+	typ errorType
+	err error
+}
+
+func (e *apiError) Error() string {
+	return fmt.Sprintf("%s: %s", e.typ, e.err)
+}
+
+var corsHeaders = map[string]string{
+	"Access-Control-Allow-Headers":  "Accept, Authorization, Content-Type, Origin",
+	"Access-Control-Allow-Methods":  "GET, POST, DELETE, OPTIONS",
+	"Access-Control-Allow-Origin":   "*",
+	"Access-Control-Expose-Headers": "Date",
+	"Cache-Control":                 "no-cache, no-store, must-revalidate",
+}
+
+// Enables cross-site script calls.
+func setCORS(w http.ResponseWriter) {
+	for h, v := range corsHeaders {
+		w.Header().Set(h, v)
+	}
+}
+
+// API provides registration of handlers for API routes.
+type API struct {
+	logger      log.Logger
+	MetricStore storage.MetricStore
+	Flags       map[string]string
+	StartTime   time.Time
+	BuildInfo   map[string]string
+}
+
+// New returns a new API.
+// logger can be nil and if log.logger is set to nil no logging is possible
+// MetricStore cannot be left empty
+// Flags is information of flags provide at initialization of pushgateway
+// BuildInfo consist of different information which are set at build time
+// StartTime is time of initialization of APIs
+func New(
+	l log.Logger,
+	ms storage.MetricStore,
+	flags map[string]string,
+	buildInfo map[string]string,
+) *API {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	return &API{
+		StartTime:   time.Now(),
+		logger:      l,
+		MetricStore: ms,
+		Flags:       flags,
+		BuildInfo:   buildInfo,
+	}
+}
+
+// Register registers the API handlers under their correct routes
+// in the given router.
+func (api *API) Register(r *route.Router) {
+	wrap := func(f http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			setCORS(w)
+			f(w, r)
+		})
+	}
+
+	r.Options("/*path", wrap(func(w http.ResponseWriter, r *http.Request) {}))
+
+	r.Get("/status", wrap(api.status))
+	r.Get("/metrics", wrap(api.metrics))
+}
+
+type metrics struct {
+	Timestamp time.Time     `json:"time_stamp"`
+	Type      string        `json:"type"`
+	Help      string        `json:"help"`
+	Metrics   []*dto.Metric `json:"metrics"`
+}
+
+func (api *API) metrics(w http.ResponseWriter, r *http.Request) {
+	familyMaps := api.MetricStore.GetMetricFamiliesMap()
+	res := make([]interface{}, 0)
+	for _, v := range familyMaps {
+		metricResponse := make(map[string]interface{})
+		metricResponse["label"] = v.Labels
+		metricResponse["last_success"] = v.LastPushSuccess()
+		for name, metricValues := range v.Metrics {
+			uniqueMetrics := metrics{
+				Type:      metricValues.GobbableMetricFamily.Type.String(),
+				Help:      *metricValues.GobbableMetricFamily.Help,
+				Timestamp: metricValues.Timestamp,
+				Metrics:   metricValues.GobbableMetricFamily.Metric,
```

Comment:
I'm afraid we cannot just wing the `dto.Metric` unmarshalling. You can easily try it out. Push the following to the Pushgateway:

```
echo 'some_metric{foo="bar",instance="bla"} NaN' | curl --data-binary @- http://localhost:9091/metrics/job/hard
```

Then `curl http://localhost:9091/api/v1/metrics` will return nothing, and the PGW will log:
```
level=error ts=2020-02-16T21:15:56.071Z caller=api.go:181 msg="error marshaling JSON" err="json: unsupported value: NaN"
```

I'm afraid you have to somehow hook into the unmarshaling and make sure sample values are rendered as strings. Perhaps you have to just do the whole marshalling manually. See how it is done in https://github.com/prometheus/prom2json .

Perhaps even just import prom2json as a library and use https://github.com/prometheus/prom2json/blob/master/prom2json.go#L65 ?

Not saying that's the best idea, just brainstorming…

Reply:
In different news: The marshaling error should have resulted in a 500 returned to the HTTP caller, but my `curl` saw a 200 with an empty response.

Reply:
> Not saying that's the best idea, just brainstorming…

Not a bad idea but unnecessary files from prom2json will be imported.

Reply:
```json
{
  "status": "success",
  "data": [
    {
      "labels": {
        "job": "hard"
      },
      "last_push_successful": true,
      "push_failure_time_seconds": {
        "time_stamp": "2020-02-17T12:12:04.796561905+05:30",
        "type": "GAUGE",
        "help": "Last Unix time when changing this group in the Pushgateway failed.",
        "metrics": [
          {
            "labels": {
              "instance": "",
              "job": "hard"
            },
            "value": "0"
          }
        ]
      },
      "push_time_seconds": {
        "time_stamp": "2020-02-17T12:12:04.796561905+05:30",
        "type": "GAUGE",
        "help": "Last Unix time when changing this group in the Pushgateway succeeded.",
        "metrics": [
          {
            "labels": {
              "instance": "",
              "job": "hard"
            },
            "value": "1.581921724796562e+09"
          }
        ]
      },
      "some_metric": {
        "time_stamp": "2020-02-17T12:12:04.796561905+05:30",
        "type": "UNTYPED",
        "metrics": [
          {
            "labels": {
              "foo": "bar",
              "instance": "bla",
              "job": "hard"
            },
            "value": "NaN"
          }
        ]
      }
    }
  ]
}
```

Reply:
@beorn7 Value is now converted into to string.